drivers/input/uinput.c

```diff
@@ -73,30 +154,173 @@ static void uinput_button_enable(FAR const struct btn_lowerhalf_s *lower,
+
+static void uinput_rpmsg_device_created(FAR struct rpmsg_device *rdev,
+                                        FAR void *priv)
 {
-  FAR const struct touch_sample_s *sample;
+  int  ret;
+  char rpmsg_ept_name[RPMSG_NAME_SIZE];
+  FAR struct uinput_rpmsg_ept_s *ept;
+  FAR struct uinput_context_s   *ctx  = (FAR struct uinput_context_s *)priv;
+  FAR struct list_node          *list = &ctx->eptlist;
+
+  ept = kmm_zalloc(sizeof(struct uinput_rpmsg_ept_s));
+  if (!ept)
+    {
+      ierr("Failed to alloc memory\n");
+      return;
+    }
 
-  sample = (FAR const struct touch_sample_s *)buffer;
-  if (sample == NULL)
+  ept->ept.priv = ctx;
+  snprintf(rpmsg_ept_name, RPMSG_NAME_SIZE,
+           RPMSG_UINPUT_NAME, ctx->name);
+  ret = rpmsg_create_ept(&ept->ept, rdev, rpmsg_ept_name,
+                         RPMSG_ADDR_ANY, RPMSG_ADDR_ANY,
+                         uinput_rpmsg_ept_cb, NULL);
+  if (ret < 0)
     {
-      return -EINVAL;
+      ierr("uinput rpmsg create failure, %s\n", rpmsg_get_cpuname(rdev));
+      kmm_free(ept);
+      return;
+    }
+
+  ept->rdev = rdev;
+  list_add_tail(list, &ept->node);
+}
+
+/****************************************************************************
+ * uinput_rpmsg_device_destroy
+ ****************************************************************************/
+
+static void uinput_rpmsg_device_destroy(FAR struct rpmsg_device *rdev,
+                                        FAR void *priv)
+{
+  FAR struct uinput_context_s   *ctx  = (FAR struct uinput_context_s *)priv;
+  FAR struct list_node          *list = &ctx->eptlist;
+  FAR struct uinput_rpmsg_ept_s *ept;
+
+  list_for_every_entry(list, ept, struct uinput_rpmsg_ept_s, node)
+    {
+      if (ept->rdev == rdev)
+        {
+          list_delete(&ept->node);
+          rpmsg_destroy_ept(priv);
+          kmm_free(ept);
+          return;
+        }
+    }
+}
+
+/****************************************************************************
+ * Name: uinput_rpmsg_initialize
+ ****************************************************************************/
+
+static int uinput_rpmsg_initialize(FAR struct uinput_context_s *ctx,
+                                   FAR const char *name)
+{
+  list_initialize(&ctx->eptlist);
+  strlcpy(ctx->name, name, UINPUT_NAME_SIZE);
+  return rpmsg_register_callback(ctx, uinput_rpmsg_device_created,
+                                 uinput_rpmsg_device_destroy, NULL);
+}
+
+/****************************************************************************
+ * Name: uinput_rpmsg_notify
+ ****************************************************************************/
+
+static void uinput_rpmsg_notify(FAR struct uinput_context_s *ctx,
+                                FAR const char *buffer, size_t buflen)
+{
+  int ret;
+  FAR struct uinput_rpmsg_ept_s *ept;
+
+  list_for_every_entry(&ctx->eptlist, ept, struct uinput_rpmsg_ept_s, node)
+    {
+      if (!is_rpmsg_ept_ready(&ept->ept))
```

Comment:
ditto

Reply:
If we can change condition to eliminate `continue` that would be good

Reply:
done